src/Functions/splitByRegexp.cpp

```diff
@@ -148,11 +150,59 @@ class SplitByRegexpImpl
 
 using FunctionSplitByRegexp = FunctionTokens<SplitByRegexpImpl>;
 
+/// Fallback splitByRegexp to splitByChar when its 1st argument is a trivial char for better performance
+class SplitByRegexpOverloadResolver : public IFunctionOverloadResolver
+{
+public:
+    static constexpr auto name = "splitByRegexp";
+    static FunctionOverloadResolverPtr create(ContextPtr context) { return std::make_unique<SplitByRegexpOverloadResolver>(context); }
+
+    explicit SplitByRegexpOverloadResolver(ContextPtr context_) : context(context_), split_by_regexp(FunctionSplitByRegexp::create(context))
+    {
+    }
+
+    String getName() const override { return name; }
+    size_t getNumberOfArguments() const override { return SplitByRegexpImpl::getNumberOfArguments(); }
+    bool isVariadic() const override { return SplitByRegexpImpl::isVariadic(); }
+
+    FunctionBasePtr buildImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr & return_type) const override
+    {
+        bool should_fallback_to_split_by_char = false;
+        const ColumnConst * col = checkAndGetColumnConstStringOrFixedString(arguments[0].column.get());
+        if (!col)
+            throw Exception(
+                ErrorCodes::ILLEGAL_COLUMN,
+                "Illegal column {} of first argument of function {}. "
+                "Must be constant string.",
+                arguments[0].column->getName(),
+                getName());
+
+        String pattern = col->getValue<String>();
+        if (pattern.size() == 1 && regex_symbols.count(pattern[0]) == 0)
+            should_fallback_to_split_by_char = true;
+
+        if (should_fallback_to_split_by_char)
+            return FunctionFactory::instance().getImpl("splitByChar", context)->build(arguments);
+        else
+            return std::make_unique<FunctionToFunctionBaseAdaptor>(
+                split_by_regexp, collections::map<DataTypes>(arguments, [](const auto & elem) { return elem.type; }), return_type);
+    }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        return split_by_regexp->getReturnTypeImpl(arguments);
+    }
+
+private:
+    ContextPtr context;
+    FunctionPtr split_by_regexp;
+    inline static const std::unordered_set<char> regex_symbols = {'^', '$', '\\', '.', '*', '+', '?', '(', ')', '[', ']', '{', '}', '|'};
```

Comment:
1) l. 181 should use `.contains()` to check membership in the unordered_set. An array of regexp symbols would be slightly nicer 2) since the pattern size is required to be 1, we can remove regexp symbols that cannot be used standalone such as braces, parentheses, possibly also `|`, `^` etc. (please check if re2 accepts them or throws an invalid pattern error).

Reply:
I'll use OptimizedRegularExpression::analyze to decide whether to fallback to `splitByChar`, thus we don't need to consider the corner cases you mentions above.